contracts/NonFungibleToken-v2.cdc

```diff
@@ -0,0 +1,158 @@
+## The Flow Non-Fungible Token standard
+
+## `NonFungibleToken` contract interface
+
+The interface that all Non-Fungible Token contracts could conform to.
+If a user wants to deploy a new NFT contract, their contract would need
+to implement the NonFungibleToken interface.
+
+Their contract would have to follow all the rules and naming
+that the interface specifies.
+
+## `NFT` resource
+
+The core resource type that represents an NFT in the smart contract.
+
+## `Collection` Resource
+
+The resource that stores a user's NFT collection.
+It includes a few functions to allow the owner to easily
+move tokens in and out of the collection.
+
+## `Provider` and `Receiver` resource interfaces
+
+These interfaces declare functions with some pre and post conditions
+that require the Collection to follow certain naming and behavior standards.
+
+They are separate because it gives the user the ability to share a reference
+to their Collection that only exposes the fields and functions in one or more
+of the interfaces. It also gives users the ability to make custom resources
+that implement these interfaces to do various things with the tokens.
+
+By using resources and interfaces, users of NFT smart contracts can send
+and receive tokens peer-to-peer, without having to interact with a central ledger
+smart contract.
+
+To send an NFT to another user, a user would simply withdraw the NFT
+from their Collection, then call the deposit function on another user's
+Collection to complete the transfer.
+
+*/
+
+import MetadataViews from "./MetadataViews.cdc"
+
+/// The main NFT contract interface. Other NFT contracts will
+/// import and implement this interface
+///
+pub contract NonFungibleToken {
+
+    /// Interface that the NFTs have to conform to
+    ///
+    pub resource interface NFT { //: MetadataViews.Resolver {
+        /// The unique ID that each NFT has
+        pub let id: UInt64
+
+        pub fun getViews(): [Type]
+        pub fun resolveView(_ view: Type): AnyStruct?
+    }
+
+    /// Interface to mediate withdraws from the Collection
+    ///
+    pub resource interface Provider {
+        /// withdraw removes an NFT from the collection and moves it to the caller
+        pub fun withdraw(withdrawID: UInt64): @AnyResource{NFT} {
+            post {
+                result.id == withdrawID: "The ID of the withdrawn token must be the same as the requested ID"
+            }
+        }
+    }
+
+    /// Interface to mediate withdraws from the Collection
+    ///
+    pub resource interface Transferable {
+        /// withdraw removes an NFT from the collection and moves it to the caller
+        pub fun transfer(id: UInt64, receiver: Capability<&AnyResource{Receiver}>): @AnyResource{NFT}
+    }
+
+    /// Interface to mediate deposits to the Collection
+    ///
+    pub resource interface Receiver {
+
+        /// deposit takes an NFT as an argument and adds it to the Collection
+        ///
+        pub fun deposit(token: @AnyResource{NFT})
```

Comment:
post condition check with `getAcceptedTypes`

Reply:
What kind of post condition do you mean?

Reply:
@bluesign do you mean a pre check that the type of the token being deposited is part of `getAcceptedTypes`? Also not sure what post condition you'd put here

Reply:
@austinkline yeah I was thinking like that, with id check not checking type on withdrawal , maybe we can prevent that hole here.